fix(curriculum): advanced node express changes for new boilerplate

ShaunSHamilton opened the pull request:
Checklist:

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] All the files I changed are in the same world language, for example: only English changes, or only Chinese changes, etc.

Affected Issues:
Closes #38882
#38837 (will not close)
Closes #38064
Closes #36281
Closes #17820
Closes #17781
Closes #17704
Closes #36050
Closes #18103 
Closes #17765
Closes #17757 
Closes #17763 
Closes #17618

Affected PRs: #38979 (made redundant), #39076 (made redundant)

<!-- Feel free to add any additional description of changes below this line -->
These changes solve many other issues, once coupled with the new boilerplate linked: https://github.com/freeCodeCamp/boilerplate-advancednode/pull/11

Commit:
fix: add tests and steps

Commit:
add necessary changes

Commit:
edit for new boilerplate

Commit:
fix: adjust content for boilerplate merge

Commit:
add 4 passing 1 failing socketio

Commit:
fix: add socketio changes

scissorsneedfoodtoo reviewed (commented):
Just some minor things I noticed while going through the project. Could you take a look @Sky020?

scissorsneedfoodtoo commented:
Also, if you don't mind, I can push a commit to update all the Glitch links in these challenges on Repl.it.

Sky020 commented on curriculum/challenges/english/06-quality-assurance/advanced-node-and-express/serialization-of-a-user-object.english.md:
```diff
@@ -18,12 +18,14 @@ passport.serializeUser((user, done) => {
-  db.collection('users').findOne(
-    {_id: new ObjectID(id)},
+  myDB(async (client) => {
+    await client.db('database').collection('users').findOne(
+      { _id: new ObjectID(id) },
       (err, doc) => {
-        done(null, doc);
+        done(null, null);
       }
-  );
+    );
+  });
```
Thank you, for pointing this out. This was an overlook of mine, as I did update the connection method, but forgot to change the lesson content.

scissorsneedfoodtoo replied:
Sorry, I could have been more clear.

In my testing, I ran into problems with the following

```js
passport.deserializeUser((id, done) => {
  myDB(async (client) => {
    await client.db('database').collection('users').findOne(
      { _id: new ObjectID(id) },
      (err, doc) => {
        done(null, doc);
      }
    );
  });
});
```

after the Clean Up Your Project with Modules challenge. I would get the error `myDataBase is not a function` (renamed from `myDB` to `myDataBase` if going by the params in the instructions for the Clean Up Your Project with Modules challenge).

In the end I had to rewrite it as follows:

```js
passport.deserializeUser((id, done) => {
  myDataBase.findOne(
    {_id: new ObjectID(id)},
      (err, doc) => {
        done(null, doc);
      }
  );
});
```

Not sure if it's a mistake somewhere I made in the project, but my suggestion is just to have it as the following in this step to prevent possible errors in later challenges:

```js
passport.deserializeUser((id, done) => {
  myDB.findOne(
    {_id: new ObjectID(id)},
      (err, doc) => {
        done(null, doc);
      }
  );
});
```

Sky020 replied:
>after the Clean Up Your Project with Modules challenge.

Sorry, Kris. I still do not understand. [This lesson (the next one)](https://github.com/freeCodeCamp/freeCodeCamp/pull/39080/files#diff-e2fa44d297bf8e4f5ef85d88109b3fd5) expects:
>Be sure to remove the <code>myDB</code> code in deserializeUser and edit your <code>done(null, null)</code> to include the <code>doc</code>.

This is all done long before the _Clean Up Your Project_ lesson. If you take a look at the Repl.it, it should be similar. So, If you think this section is unclear as to what to do, then we definitely need to reword the instructions.

scissorsneedfoodtoo replied:
Hi @Sky020 @Sky020, sorry, I must be missing something in these two steps.

Just to clarify, in Serialization of a User Object, learners will add the following to `server.js`:

```js
passport.serializeUser((user, done) => {
  done(null, user._id);
});
passport.deserializeUser((id, done) => {
  myDB(async client => {
    await client
      .db('database')
      .collection('users')
      .findOne({ _id: new ObjectID(id) }, (err, doc) => {
        done(null, null);
      });
  });
});
```

And they will comment out everything in `deserializeUser` except for `done(null, null);` so it doesn't throw an error.

Then in the next challenge, Implement the Serialization of a Passport User, learners set up their `MONGO_URI` string in `.env`, and encompass the code in what's provided in the description.

Also I see where they're supposed to refactor `deserializeUser`, which I must have missed the first time around. The only thing I did while running through the project before was to change `done(null, null);` to `done(null, doc);`.

So the current instructions, "Be sure to remove the `myDB` code in `deserializeUser` and edit your `done(null, null)` to include the doc", means to go from this in the previous challenge,

```js
passport.deserializeUser((id, done) => {
//   myDB(async client => {
//     await client
//       .db('database')
//       .collection('users')
//       .findOne({ _id: new ObjectID(id) }, (err, doc) => {
        done(null, null);
//       });
//   });
});
```

To this in Implement the Serialization of a Passport User:

```js
passport.deserializeUser((id, done) => {
  myDataBase.findOne({ _id: new ObjectID(id) }, (err, doc) => {
    if (err) return console.error(err);
      done(null, doc);
    });
});
```

Is that correct?

Sky020 replied:
@scissorsneedfoodtoo , yes that looks correct.

scissorsneedfoodtoo replied:
@Sky020, great, thanks for confirming that.

I think it might be a bit too much of a jump between the two steps. "Be sure to remove the `myDB` code in `deserializeUser` and edit your `done(null, null)` to include the `doc`" almost sounds like learners are supposed to remove everything in `deserializeUser` except for `done(null, null)`, which they change to `done(null, doc)`.

One way we can ease the transition a bit is to change the example code in Serialization of a User Object to something like:

```js
passport.serializeUser((user, done) => {
  done(null, user._id);
});

passport.deserializeUser((id, done) => {
  myDB.findOne({ _id: new ObjectID(id) }, (err, doc) => {
    done(null, null);
  });
});
```

with the expectation they'll comment out everything but `done(null, null);`. None of the instructions should have to change in this step.

Then in the next step Implement the Serialization of a Passport User, we could modify the second to last sentence to read "Be sure to change `myDB` to `myDataBase` in `deserializeUser` and edit your `done(null, null)` to include the `doc`."

Does that sound like a good change? Or is there another way you'd like to break these steps up?

Sky020 replied:
I have done something similar; I decided not to have `myDB`, if it will be commented out and renamed, anyway.

scissorsneedfoodtoo replied:
Awesome, thanks for making the changes. LGTM :+1:

Sky020 commented:
> Also, if you don't mind, I can push a commit to update all the Glitch links in these challenges on Repl.it.

@scissorsneedfoodtoo , No, I do not mind. Absolutely, go for it.

We also need to discuss how to update the gists with the _code up to this point_...

Commit:
fix: update wording and http test

Co-authored-by: Kristofer Koishigawa <scissorsneedfoodtoo@gmail.com>

scissorsneedfoodtoo commented:
@Sky020, awesome, sounds great. I'll work on updating those links and can try working through the project again to get some of those Gists in place. I'll need to check which account we should have that under.

scissorsneedfoodtoo committed:
fix: replace glitch remix urls with repl.it urls

Commit:
integrate steps between lessons 4 and 5

scissorsneedfoodtoo commented on curriculum/challenges/english/06-quality-assurance/advanced-node-and-express/implement-the-serialization-of-a-passport-user.english.md:
```diff
@@ -7,53 +7,74 @@ forumTopicId: 301556
-You can now uncomment the block in deserializeUser and remove your <code>done(null, null)</code>. Be sure to set <em>DATABASE</em> in your .env file to your database's connection string (for example: <code>DATABASE=mongodb://admin:pass@mlab.com:12345/my-project</code>). You can set up a free database on <a href='https://mlab.com/welcome/'>mLab</a>. Congratulations- you've finished setting up seri ...
+Be sure to uncomment the <code>myDataBase</code> code in deserializeUser, and edit your <code>done(null, null)</code> to include the <code>doc</code>.
+
+You can set up a free database on <a href='https://mlab.com/welcome/'>mLab</a>. Congratulations- you've finished setting up serialization!
```
Sorry, I must have missed this the first time around. This can be changed to something like "You can set up a free database on "<a href='https://www.mongodb.com/cloud/atlas'>MongoDB Atlas</a>."

It might also make sense to move this sentence to line 16 where the connection string is mentioned.

Sky020 replied:
Thanks for catching that.

scissorsneedfoodtoo commented:
@RandellDawson, could you take a look at this?

Commit:
add mongodb altas link

Sky020 commented:
@scissorsneedfoodtoo and @RandellDawson . To help, I have started creating the gists: https://gist.github.com/Sky020

It is just easier this way for you to QA, and work off of. I should have them all finished, soon.

scissorsneedfoodtoo commented:
Thanks @Sky020, that's a huge. I started on those last week, but only got to the first few of checkpoints (Set up Passport, and Implement the Serialization of a Passport User).

Are you thinking of having a link to a Gist in each challenge? Either way is totally fine, it would just be a bit more to change if we make modifications in the future.

Either way, if you'd like to continue creating the Gists, I can start adding them to one of our accounts and start updating the links.

Sky020 commented:
@scissorsneedfoodtoo 
>Are you thinking of having a link to a Gist in each challenge?

I will leave this up to you. I am creating one for each, but I can fully understand some of the lessons are simple/laid-out enough that they do not require one.

Just a heads-up, I am busy changing the _Registration of New Users_ lesson tests, to mimic those of the _Exercise Tracker_ project, so that users do not need to delete the database record, every time they want to run the tests.

Sky020 commented:
@scissorsneedfoodtoo Before I commit these test changes, does this look reasonable?
```
tests:
  - text: You should register route and display on home.
    testString: 'getUserInput => $.get(getUserInput(''url'')+ ''/_api/server.js'') .then(data => { assert.match(data, /showRegistration:( |)true/gi, ''You should be passing the variable "showRegistration" as true to your render function for the homepage''); assert.match(data, /register[^]*post[^]*findOne[^]*username:( |)req.body.username/gi, ''You should have a route accepted a post request on register that querys the db with findone and the query being "username: req.body.username"''); }, xhr => { throw new Error(xhr.statusText); })'
  - text: Registering should work.
    testString: "async getUserInput => {
      const url = getUserInput('url');
      const user = `freeCodeCampTester${Date.now()}`;
      const xhttp=new XMLHttpRequest();
      xhttp.onreadystatechange = function() {
        if (this.readyState == 4 && this.status == 200) {
          test(this);
        } else {
          throw new Error(`${this.status} ${this.statusText}`);
        }
      };
      xhttp.open('POST', url+'/register', true);
      xhttp.setRequestHeader('Content-type', 'application/x-www-form-urlencoded');
      xhttp.send(`username=${user}&password=${user}`);  

      function test(xhttpRes) {
        const data = xhttpRes.responseText;
        assert.match(data, /Profile/gi, 'Register should work, and redirect successfully to the profile.');
      }
    }
    "
  - text: Login should work.
    testString: "async getUserInput => {
      const url = getUserInput('url');
      const user = `freeCodeCampTester${Date.now()}`;
      const xhttpReg = new XMLHttpRequest();
      xhttpReg.onreadystatechange = function() {
        if (this.readyState == 4 && this.status == 200) {
          login();
        } else {
          throw new Error(`${this.status} ${this.statusText}`);
        }
      };
      xhttpReg.open('POST', url+'/register', true);
      xhttpReg.setRequestHeader('Content-type', 'application/x-www-form-urlencoded');
      xhttpReg.send(`username=${user}&password=${user}`);

      function login() {
        const xhttp=new XMLHttpRequest();
        xhttp.onreadystatechange = function() {
          if (this.readyState == 4 && this.status == 200) {
            test(this);
          } else {
            throw new Error(`${this.status} ${this.statusText}`);
          }
        };
        xhttp.open('POST', url+'/login', true);
        xhttp.setRequestHeader('Content-type', 'application/x-www-form-urlencoded');
        xhttp.send(`username=${user}&password=${user}`);  
      }
      function test(xhttpRes) {
        const data = xhttpRes.responseText;
        assert.match(data, /Profile/gi, 'Login should work if previous test was done successfully and redirect successfully to the profile.');
        assert.match(data, new RegExp(user, 'g'), 'The profile should properly display the welcome to the user logged in');
      }
    }
    "
```

scissorsneedfoodtoo commented:
@Sky020, awesome! Looks great. I was also playing around with that yesterday but ran into some CORS issues with `fetch`. Your changes look good to me. Go ahead and push them up and I'll test that challenge out locally.

About the Gists, it wouldn't hurt to have one for each challenge if you're already in the process. I'm sure anyone who has trouble with a step would appreciate the extra help, and we could always remove some for the very easy challenges later.

Commit:
edit test to not require db deletion

Sky020 commented:
@scissorsneedfoodtoo **AN IMPORTANT POINT**: We need to include somewhere within these lessons:
- If a user is using a tool like Repl.it, the authentication steps will not work within the PIP browser. The new window must be used.

I spent the last two days banging my head against the wall, as to why this _all-of-a-sudden_ was not working.

I have pushed my test changes. The issue I am trying to debug is: I am not throwing the correct assertion fail, as the tests pass, but the console says otherwise:
![image](https://user-images.githubusercontent.com/51722130/90161176-add71200-dd8a-11ea-8270-400daea77902.png)

Commit:
correct register routing and formatting

Commit:
fix typos and formatting

Sky020 commented:
The Gists are done. The issue above still remains.

Here are some final points which might need discussing:
1) The above mentioned issue...
2) Socket.io is only set-up to work with the _GitHub_ auth routes. Should we refactor in any way to include the `/login` and `/register` routes. **NOTE**: This is not trivial, as Socket.io is setup to expect the `user` object from GitHub, which is structured differently to the `user` object from the app.
3) One or two of the lessons require quite a lot of code to be added. Should/could this be spread between other lessons?

scissorsneedfoodtoo commented:
@Sky020, awesome, thank you. I'll create a fork of your Gist for each challenge under the CamperBot account and push a commit to your PR with a link at the bottom of each description.

For the issue with the PIP browser, do you remember the challenge where you first ran into that issue? It might be enough to include some text in that challenge about having to open up a new window. It might not hurt to include a similar note in the next challenge or two as a reminder.

About the assertion errors, I can look into that.

For the other issues you mentioned, it's probably fine to have Socket.io only work with the GitHub auth for now. That's how the original project was structured, and it would take a bit too long to refactor the project now. While it might be something to consider in the future, you've already done a lot of great work to simplify the project. 

I don't think we need to hold up merging this to get Socket.io working for those other routes now. Same goes for spreading the code out between more lessons. That could be tackled in another PR further down the road.

scissorsneedfoodtoo committed:
fix: typos, standardize spacing, and remove unnecessary hr elements

scissorsneedfoodtoo committed:
fix: add/update links

Add or update Gist solution links at the bottom of each challenge. Also add a missing link/text to the top of one of the challenges.

Sky020 commented:
@scissorsneedfoodtoo 
> For the issue with the PIP browser, do you remember the challenge where you first ran into that issue?

I think it was the _Registration of New Users_ challenge. There is something about the way the routing is done between the server and the in-editor frame which prevents the session from persisting. I do not know much about the topic, but the solution was to not use it from this point forward.

scissorsneedfoodtoo commented:
Okay, thanks @Sky020. I can take a crack at adding some text to the challenge, and maybe to the intro page that mentions the PIP browser might have some issues in some challenges, and to open the project up in a separate window. I'll go ahead and fix the merge conflicts before adding that.

Another thing @RandellDawson and I talked about is removing the "As a reminder, this project is being built upon the following starter project" after the first challenge. Our thinking is that, while it's good to have a reminder, if someone takes a break and decides to pick up the project later, they'll just create a new project on Repl.it if they click that link. 

Instead, we'll just have those links on the intro page and in the description for the first challenge.

Sky020 commented:
> Instead, we'll just have those links on the intro page and in the description for the first challenge.

@scissorsneedfoodtoo That makes sense. Do you want me to go ahead and remove that section, or will you?

scissorsneedfoodtoo committed:
fix: remove Repl.it/boilerplate repo links from all but first challenge

scissorsneedfoodtoo committed:
fix: add target='_blank' to links in challenges

scissorsneedfoodtoo commented:
Sorry @Sky020, was deep in the command line trying to resolve those merge conflicts. I went ahead and removed all that text from everything but the first challenge. The intro page already has links to start the project on Repl.it and to the boilerplate, so no need to worry about that.

I didn't get a chance to add the text about the PIP browser. It would be great if you have some time to add that. If not, no worries. I'll pick that up in the morning.

Commit:
add note about PIP browser issues

Sky020 commented:
Thank you, for sorting out those conflicts, Kris. I have pushed the following not about the PIP browser:
![image](https://user-images.githubusercontent.com/51722130/90651390-de102c00-e234-11ea-908c-5871780c7c38.png)

By all means, change the wording and style.

scissorsneedfoodtoo commented:
@Sky020, awesome, thanks for adding that. Just took a look and it reads well. My only suggestion would be to move the text to the end of that description section and remove the `hr` element. A lot of current challenges like [Perform New Updates on a Document Using model.findOneAndUpdate()](https://www.freecodecamp.org/learn/apis-and-microservices/mongodb-and-mongoose/perform-new-updates-on-a-document-using-model-findoneandupdate) have the note at the end of the description/instructions.

Yes, I did remove most if not all of the existing `hr` elements. I thought it was a bit jarring that they were used so often in the later challenges. Ideally these challenges would be written in such a way that there wouldn't need to be any extra `hr` elements to split up in the information. There would just be the description and instructions, which are both separated by an `hr`.

But with the way a lot of backend challenges are written, one step could require a bunch of code changes, with instructions spread throughout the description. In cases like that, I could see there being just one longer description section.

Commit:
move PIP note to end of instructions

scissorsneedfoodtoo commented:
@RandellDawson, this PR is ready for review when you get a chance.

moT01 reviewed (commented):
Very nicely done @Sky020 🎉 and @scissorsneedfoodtoo 🎉 

I was able to go through the whole section. Since this is such a big improvement, I am not going to complain about grammar or instructions - most of what I didn't like there was preexisting and can be improved upon with future PR's.

I think I'm going to vote to just add this. Note that I would wait on merging the boilerplate PR until this is pushed to production.

A few thoughts that I did have...
- Minor - There's some indentation and spacing issues in a number of the gists - as well as some comments that could probably be removed.

- Minor - The `sample.env` could just be `.env` so users don't have to copy it? (this one is on the boilerplate)

- This is the only one worth discussing I think - Wondering if we should try and lock all the dependencies to a specific version so if one of them makes a change that would break something on this - it won't break.

edit: also would be nice to make sure nodemon is working so users don't have to restart the server

Sky020 commented:
@moT01 
>Minor - There's some indentation and spacing issues in a number of the gists - as well as some comments that could probably be removed.

Do you mean this type of comment:
```
//Change the response to render the Pug template
```
That is one of the only comments I would remove.

> Minor - The sample.env could just be .env so users don't have to copy it? (this one is on the boilerplate)

I was unsure what would happen, if I had `.env` in `.gitignore` and wanted to push `.env` to the repo without changing `.gitignore`. I think it is best leaving `sample.env` as is, to prevent eager campers from pushing their secrets everywhere.

>  Wondering if we should try and lock all the dependencies to a specific version

Similar point of discussion: Would creating a section/file/addition-to-readme with all of the current (right now) versions of the packages used, be beneficial anyway? A simple _copy-pasta_ of the package.json as it is completed.

moT01 commented:
I was referring to [this comment](https://gist.github.com/camperbot/bf95a0f74b756cf0771cd62c087b8286#file-client-js-L16) - perhaps it's there for a reason - and yes, the two space indentation to four space indentation you can see in [these files](https://gist.github.com/camperbot/2d06ac5c7d850d8cf073d2c2c794cc92) - not a big deal - no need to change it, I was just pointing it out.

> I was unsure what would happen, if I had .env in .gitignore and wanted to push .env to the repo without changing .gitignore. I think it is best leaving sample.env as is, to prevent eager campers from pushing their secrets everywhere.

Perhaps that's a good idea.

> Similar point of discussion: Would creating a section/file/addition-to-readme with all of the current (right now) versions of the packages used, be beneficial anyway? A simple copy-pasta of the package.json as it is completed.

That could be beneficial I suppose - I was thinking to adjust the instructions to tell users to use specific versions. Instead of "install pug" - "install pug version 3.0.0" - but I'm not sure it's worth it to make all those adjustments at this point. I suppose you could possibly just put them all in the package.json file and then all users have to do is require them - that doesn't sound too bad.

Sky020 commented:
@scissorsneedfoodtoo and @RandellDawson I would appreciate you weighing in on the switch to including all necessary packages in the boilerplate `package.json`? It would involve removing lines in the lessons about _"install package x and y"_

RandellDawson commented:
> @scissorsneedfoodtoo and @RandellDawson I would appreciate you weighing in on the switch to including all necessary packages in the boilerplate `package.json`? It would involve removing lines in the lessons about _"install package x and y"_

I am still reviewing both this PR and the corresponding boilerplate PR.  Yesterday, I mentioned to you that mongodb should be apart of the boilerplate's `package.json` file, because if the user just starts the project in development mode, they would get an error right off the bat.  My first thought is that the only packages that need to be present in the boilerplate, are ones that would prevent an error in the terminal when ran locally (npm run dev) the first time.  After that, it would be fine to just have the user `install package x and y` in the various challenges.

RandellDawson reviewed (dismissed):
@Sky020 There are many issues with tests that have nothing to do with your changes, but will need to be handled in a later PR. For example, in the Clean Up Your Project challenge, you can pass to the next challenge without adding the dependencies to auth.js and routes.js

One of the only things I would recommend fixing in a separate PR is making it clear in the Setup the Environment challenge, that client.js is in the public folder. @scissorsneedfoodtoo added a comment in the client.js file on the gist for that challenge that explains where the file is located, but I think the instructions should indicate this.

I am fine approving this PR "as-is" and leaving the other changes regarding how to word the package versions for a separate PR.  I will leave a separate review on the boilerplate PR.

scissorsneedfoodtoo committed:
Merge branch 'master' into fix/advanced-node-express

scissorsneedfoodtoo commented:
Fixed a merge conflict we inadvertently caused in `serialization-of-a-user-object.english.md`.

moT01 reviewed (approved):
LGTM 🎉 Thanks for all the hard work on this @Sky020 🎉 and reviewers 🎉

scissorsneedfoodtoo commented:
Thanks again for all of the hard work you put into this @Sky020! Looking forward to seeing this live :+1: :+1: 

@moT01, thanks for catching the issue with the comments and spacing on the Gists. I can fix those from the Camperbot account.
